operator/duties/scheduler.go

```diff
@@ -0,0 +1,360 @@
+}
+
+// ValidatorController represents the component that controls validators via the scheduler
+type ValidatorController interface {
+	ActiveValidatorIndices(logger *zap.Logger, epoch phase0.Epoch) []phase0.ValidatorIndex
+	GetOperatorShares() []*types.SSVShare
+}
+
+type ExecuteDutyFunc func(logger *zap.Logger, duty *spectypes.Duty)
+
+type SchedulerOptions struct {
+	Ctx                 context.Context
+	BeaconNode          BeaconNode
+	Network             networkconfig.NetworkConfig
+	ValidatorController ValidatorController
+	ExecuteDuty         ExecuteDutyFunc
+	IndicesChg          chan struct{}
+	Ticker              SlotTicker
+	BuilderProposals    bool
+}
+
+type Scheduler struct {
+	beaconNode          BeaconNode
+	network             networkconfig.NetworkConfig
+	validatorController ValidatorController
+	slotTicker          SlotTicker
+	executeDuty         ExecuteDutyFunc
+	builderProposals    bool
+
+	handlers            []dutyHandler
+	blockPropagateDelay time.Duration
+
+	reorg      chan ReorgEvent
+	indicesChg chan struct{}
+	ticker     chan phase0.Slot
+	waitCond   *sync.Cond
+	pool       *pool.ContextPool
+
+	headSlot                  phase0.Slot
+	lastBlockEpoch            phase0.Epoch
+	currentDutyDependentRoot  phase0.Root
+	previousDutyDependentRoot phase0.Root
+}
+
+func NewScheduler(opts *SchedulerOptions) *Scheduler {
+	s := &Scheduler{
+		beaconNode:          opts.BeaconNode,
+		network:             opts.Network,
+		slotTicker:          opts.Ticker,
+		executeDuty:         opts.ExecuteDuty,
+		validatorController: opts.ValidatorController,
+		builderProposals:    opts.BuilderProposals,
+		indicesChg:          opts.IndicesChg,
+		blockPropagateDelay: blockPropagationDelay,
+
+		handlers: []dutyHandler{
+			NewAttesterHandler(),
+			NewProposerHandler(),
+			NewSyncCommitteeHandler(),
+		},
+
+		ticker:   make(chan phase0.Slot),
+		reorg:    make(chan ReorgEvent),
+		waitCond: sync.NewCond(&sync.Mutex{}),
+	}
+	if s.builderProposals {
+		s.handlers = append(s.handlers, NewValidatorRegistrationHandler())
+	}
+	return s
+}
+
+type ReorgEvent struct {
+	Slot     phase0.Slot
+	Previous bool
+	Current  bool
+}
+
+func (s *Scheduler) Start(ctx context.Context, logger *zap.Logger) error {
+	logger = logger.Named(logging.NameDutyScheduler)
+	logger.Info("duty scheduler started")
+
+	// Subscribe to head events.  This allows us to go early for attestations & sync committees if a block arrives, as well as
+	// re-request duties if there is a change in beacon block.
+	s.pool = pool.New().WithContext(ctx).WithCancelOnError()
+	subscriptionCtx, subscribed := context.WithCancel(ctx)
+	s.pool.Go(func(ctx context.Context) error {
+		defer subscribed()
+		err := s.beaconNode.Events(ctx, []string{"head"}, s.HandleHeadEvent(logger))
+		if err != nil {
+			return fmt.Errorf("failed to subscribe to head events: %w", err)
+		}
+		return nil
+	})
```

Comment:
what if an error is returned? subscribed is still called and we act like we have head events. I think there's no value in this code being inside a goroutine, it just adds complexity.

Reply:
see [only call subscribed if there was no error](https://github.com/bloxapp/ssv/pull/1044/commits/0ed04167602bfd2f7592b0ada5fd6357aff7b35a)

Reply:
There is a convention in golang to use the Start function to initiate an operation in a non-blocking manner, often launching goroutines to do the work in the background.

the subscription to head events is running in a goroutine because it needs to be non-blocking. If the subscription were moved out of the goroutine, it could potentially block the rest of the Start function, which goes against the convention.

Reply:
This convention adheres to long actions. Subscribing to headevents should be a short action, and since we don't want to continue if we can't subscribe, it's better to do it in a synchronous matter. After [only call subscribed if there was no error](https://github.com/bloxapp/ssv/pull/1044/commits/0ed04167602bfd2f7592b0ada5fd6357aff7b35a), the situation is now worse, if there was an error all our handler go routines will hang waiting to the subscribe context to be done but it won't event be done, the handler will never run.
When you write a start function, it's best to run in the main start go routine stuff that are mandatory for the function to run, then return error if they fail, even if it blocks for a short time. after that you can start any long running tasks in go routine and unblock.

Reply:
See this exmaple for a start function

```golang

func Start(input string) error {

    if input != GoodInput {
        return ErrBadInput
    }
    if err := CheckRequiredService(); err != nil {
        return err
    }

    go ServiceEventLoop()

    return nil

}

Reply:
> the situation is now worse, if there was an error all our handler go routines will hang waiting to the subscribe context to be done but it won't ever be done, the handler will never run.

if there is an error subscribing 
`s.pool = pool.New().WithContext(ctx).WithCancelOnError()`
should cancel the context on error 

and the `Wait`
```
if err := n.dutyScheduler.Wait(); err != nil {
  logger.Fatal("duty scheduler exited with error", zap.Error(err))
}
```
should catch it